#29 history - 후회여부 저장,  type으로 계산하던 코드 cost 입력되는 금액으로 계산할수 있도록 변경

twonabi opened the pull request:
domain 에서 Transaction Type, Type 삭제.

twonabi committed:
refactor(#29): data.sql Transaction Type 제거 테스트를 위한 history의 querydsl 수정

twonabi committed:
feat, refactor(#29): History엔티티.
- 후회 여부 저장 기능 추가
- Transaction Type 제거 및 수정

twonabi committed:
feat, refactor(#29): History 컨트롤러.
- 후회 여부 저장 기능 추가
- Transaction Type 제거 및 수정

twonabi committed:
feat, refactor(#29): History repository.
- 지출, 수입 금액 계산식 변경(type-> cost)

twonabi committed:
feat, refactor(#29): History Response.
- Transaction Type 제거

twonabi committed:
feat, refactor(#29): History엔티티.
- 후회 여부 저장 기능 추가
- 히스토리 목록과 총 지출, 총 수입 포함을 위해Transaction Type 타입 제거 후 반환 방식 변경.( ist<HistoryResponse>  -> Map<String, Object> )

twonabi committed:
refactor(#29): 사용되지 않을 TransactionType, Type 삭제

twonabi committed:
refactor(#29): HistoryController 사용되지 않던 코드 변경

twonabi committed:
refactor(#29): HistoryController 사용되지 않던 import문 삭제

re-mora commented on porko-service/src/main/java/io/porko/history/service/HistoryService.java:
```diff
@@ -22,38 +24,53 @@ public class HistoryService {
+    @Transactional
+    public HistoryResponse updateRegretStatus(Long historyId, Boolean regret) {
+        History history = historyRepo.findById(historyId)
+                .orElseThrow(() -> new RuntimeException("상세조회를 찾을 수 없습니다: " + historyId));
```
RuntimeException 보다는 CustomException class와 ErrorCode를 구성해서 해당 도메인에서 다루는 예외를 조금 더 상세히 처리 할 수 있도록 해보는건 어떨까요?

re-mora commented on porko-service/src/main/java/io/porko/history/controller/HistoryController.java:
```diff
@@ -2,37 +2,46 @@
-    @GetMapping("history")
-    ResponseEntity<List<HistoryResponse>> getHistoryList(
+    @GetMapping
+    ResponseEntity<Map<String, Object>> getHistoryList(
```
기존 DTO 응답 방식에서 Collection 응답 방식으로 변경하신 이유가 있을까요?

re-mora commented on porko-service/src/main/java/io/porko/history/controller/HistoryController.java:
```diff
@@ -2,37 +2,46 @@
             @RequestParam(value = "startDate", required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
             @RequestParam(value = "endDate", required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate) {
 
+        Map<String, Object> response;
```
초기화 되지 않은 객체에 접근 시 해당 API가 처리해야하는 로직과는 관계없이 NPE가 발생할 수 있을 것 같아요.

re-mora commented on porko-service/src/main/java/io/porko/history/controller/model/HistoryResponse.java:
```diff
@@ -3,51 +3,42 @@
-                totalSpent,
-                totalEarned,
                 history.getRegret(),
                 null
```
null을 선언적으로 다루는 것은 주의가 많이 필요해요. 비지니스에서 null이 가지는 의미가 없다면 해당 필드를 제외한 생성자를 하나 더 추가하시는 건 어떨까요?

re-mora commented on porko-service/src/main/java/io/porko/history/service/HistoryService.java:
```diff
@@ -22,38 +24,53 @@ public class HistoryService {
     private final MemberQueryRepo memberQueryRepo;
 
-    public List<HistoryResponse> getThisMonthHistoryList(Long loginMemberId) {
+    public Map<String, Object> getThisMonthHistoryList(Long loginMemberId) {
```
이 부분도 DTO 응답 방식에서 Collection 응답 방식으로 바꾸신 이유가 있을까요?

re-mora commented on porko-service/src/main/java/io/porko/history/service/HistoryService.java:
```diff
@@ -22,38 +24,53 @@ public class HistoryService {
     }
 
-    public List<HistoryResponse> getHistoryListByDate(Long loginMemberId, LocalDate startDate, LocalDate endDate) {
+    public Map<String, Object> getHistoryListByDate(Long loginMemberId, LocalDate startDate, LocalDate endDate) {
```
여기도! DTO 응답 방식에서 Collection 응답 방식으로 변경되었네요!

re-mora commented on porko-service/src/main/java/io/porko/history/service/HistoryService.java:
```diff
@@ -22,38 +24,53 @@ public class HistoryService {
+        return HistoryResponse.ofDetail(history);
+    }
+
+    public Map<String, Object> fetchHistoryList(Long loginMemberId, LocalDate startDate, LocalDate endDate) {
```
여기도 DTO -> Collection!

re-mora reviewed (changes requested):
고생 많으셨어요, 수정을 고민해 볼만한 몇가지 코멘트를 남겨드렸으니 확인 부탁드려요 👍

twonabi committed:
refactor(#29): HistoryResponse에서 응답 방식  좀 더 구체적으로 나눴습니다.

- HistoryDetailResponse, HistoryListResponse 로 분류

twonabi committed:
refactor(#29): HistoryController 응답 방식 변경 및 정리

twonabi committed:
refactor(#29): History 엔티티 외부 접근 수준 변경 및 성능 최적화

- @NoArgsConstructor 에서 @NoArgsConstructor(access = AccessLevel.PROTECTED) 로 변경
- @ManyToOne 에서 @ManyToOne(fetch = FetchType.LAZY) 로 변경

twonabi committed:
feat(#29): History 예외 처리 추가 및 수정

twonabi committed:
feat(#29): HistoryService 예외 처리 수정 및 응답방식 변경

re-mora commented on porko-service/src/main/java/io/porko/history/controller/model/HistoryResponse.java:
```diff
@@ -13,8 +13,8 @@ public record HistoryResponse(
         String payType,
-        Boolean regret,
-        String memo
+        Boolean regret
```
null이 필요하지 않은 항목은 기본타입인 boolean을 적용해보는건 어떠세요?
추가로 regret이라는 필드 명만 보고 boolean 타입인지 유추하기 어려울 수 있을것 같아요. isRgret은 어떠세요?

re-mora commented on porko-service/src/main/java/io/porko/history/exception/HistoryErrorCode.java:
👍

re-mora commented on porko-service/src/main/java/io/porko/history/exception/HistoryException.java:
👍

re-mora reviewed (approved):
고생 많으셨습니다. 머지할게요~! 👍
